darklang/dark
# Port a lot of DB libraries and machinery

Port DB libraries:
- add a way to specific DBs in tests
- clear DB before tests with DBs in them
- add DBs (and secrets) to global variables
- Port old DB libraries (that do nothing)
- add appropriate formats for saving Dvals, and Sql convenience functions: `Sql.roundtrippableDval`, etc
- implement all of UserDB.fs
- add support for DB::set_v1, DB::get_V1, DB::query_v1, DB::query_v2, DB::query_v3, DB::getAll_v1, DB::getAllWithKeys_v1, DB::getAllWithKeys_v2, DB::getExisting_v0, DB::getManyWithKeys_v0, DB::getManyWithKeys_v1, DB::getMany_v1, DB::getMany_v2, DB::getMany_v3, DB.queryWithKey_v1, DB.queryWithKey_v2, DB::queryOne_v1, DB::queryOne_v4, DB::queryOneWithKey_v1, DB::queryOneWithKey_v2, DB::getAll_v1
- add support for other functions i needed for these tests: List::sortBy_v0, Date::parse_v2, Date::<_v0, Date::>_v0, Date::<=_v0, Date::>=_v0
- implement most of SqlCompiler, including using function signatures instead of hardcoding them in the SqlCompiler


Other changes:
- show number of commented out tests in test harness
- create ExecutionState
- add secrets to canvas
- move shortcuts to own file
- move http error checking
- fix http "Not found" tests
- create function list lazily
- get test user/canvas lazily and cache it to speed up tests
- disable HTTP logging when testing
- match EVariable interpreter implementation
- fix sha1digest 
- add Tablecloth.Map.merge
- support try/with in TaskOrValue (and fix fallout!)
- add Ast.fs with traversal functins
- move functions with no module name to LibNoModule.fs
- a few expansions to the F# parser (field access, toString, parsing Date.(<)
- fix randomString (and make it print only capital letters)
- error check when making error message for functions with wrong number of args
- troubleshooting info for VSCode
- make TestUtils.fs with shared test code
